L2/contract_upgrade_test.md

```diff
@@ -0,0 +1,201 @@
+# OP Stack Smart Contract Upgrade: v2.0.0 → v3.0.0 via OPCM
+
+## Goal
+
+We want to walk through the OP Stack smart contract upgrade workflow — specifically upgrading from `op-contracts/v2.0.0` to `op-contracts/v3.0.0` using the **OPCM** — in order to validate the following:
+
+1. **Familiarity with the upgrade workflow**, especially since we're using a customized `OptimismPortal2` contract, which may require additional steps beyond the standard flow.
+2. **Compatibility with Gnosis Safe**, ensuring the upgrade process works seamlessly when executed via a Gnosis Safe wallet.
+
+> **Important:** There is a one-to-one mapping between each OPCM and the target contracts version. You **must** deploy a dedicated OPCM for each version you plan to upgrade to. OPCM and `op-deployer` **do not support** upgrades between custom or non-standard versions.
```

Comment:
It's a bit ambiguous, OPCM is part of the contract release, it's one of the contracts released.

Reply:
Fix in [commit](https://github.com/QuarkChain/pm/pull/40/commits/0229431ba49e5310d82c7a318b94e626dcf86209)